API for server side tabular output

smarterclayton opened the pull request:
These are the APIs necessary to implement propsoal kubernetes/community#363

They consist of a new meta group (v1alpha1) that indicates these are alpha apis for the server as a whole, a new kind `TableList` which is a simple row + header arranged table capable of returning both object and columnar data, a `TableListOptions` for altering the behavior of the return, and `PartialObjectMetadata` which is an "interface" style API object which allows a client to ask any object for their metadata (without having to know how to parse the object or perform gymnastics).

Extracted from #40848

A few minor tweaks still required.

smarterclayton committed:
Add interface{} to openapi basic types

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44685)
<!-- Reviewable:end -->

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1alpha1/types.go:
```diff
@@ -0,0 +1,96 @@
+*/
+
+// package v1alpha1 is alpha objects from meta that will be introduced.
+package v1alpha1
```
Should this be v2alpha1? Or are you planning on just copying these types into v1 when they graduate?

smarterclayton replied:
The latter.

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1alpha1/types.go:
```diff
@@ -0,0 +1,96 @@
+}
+
+// TableHeader contains information about the cells returned for the object.
+type TableHeader struct {
```
This is a header for a column, not a header for a table.

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1alpha1/types.go:
```diff
@@ -0,0 +1,96 @@
+
+// Table is a tabular representation of a set of API resources. The server transforms the
+// object into a set of preferred columns for quickly reviewing the objects.
+type Table struct {
```
TableRows? TableRowList?

smarterclayton replied:
In HTML this would be a TABLE.  Should the naming here be consistent with PodList or with an HTML table?

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1alpha1/types.go:
```diff
@@ -0,0 +1,96 @@
+	Description string `json:"description"`
+}
+
+type TableItem struct {
```
Item sounds like it means cell, but this is actually a row. TableRow?

smarterclayton replied:
We use Item to represent items in a list - rather not be inconsistent.

smarterclayton replied:
I.e. this is a list of items the same way it would be in a PodList

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1alpha1/types.go:
```diff
@@ -0,0 +1,96 @@
+type TableItem struct {
+	// cells will be as wide as headers and may contain strings, numbers, booleans, simple maps, or lists, or
+	// null. See the type field of the header for a more detailed description.
+	Cells []interface{} `json:"cells"`
```
wow, not sure about `interface{}`. Can we define a more specific type, even if it is just a typedef for `interface{}`?

smarterclayton replied:
I can define one and error out if you use something out of bounds.

smarterclayton replied:
typedef is breaking in openapi - will open a follow up

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1alpha1/types.go:
```diff
@@ -0,0 +1,96 @@
+	// Standard object's metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+	// +optional
+	v1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
```
Do we want to make a version of object meta that includes only some fields?

smarterclayton replied:
I don't think it's worth it.  It's more confusing to return certain fields than to just return all metadata.

smarterclayton replied:
I.e. i'm also solving for Garbage collector here (which only needs this data).

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/openapi/common.go:
```diff
@@ -124,26 +124,27 @@ type Config struct {
+		"number":      {"number", ""},
+		"boolean":     {"boolean", ""},
+		"[]byte":      {"string", "byte"}, // base64 encoded characters
+		"interface{}": {"object", ""},
```
@mbohlool Do you have an opinion on this?

lavalamp reviewed (commented):
Looks pretty close. Mostly I am nitpicking names.

smarterclayton committed:
Add Table to meta.k8s.io/v1alpha1

Table is a generic tabular output that can be returned by servers to
enable clients to display data simply.

smarterclayton committed:
generated: Protobuf for v1alpha1

smarterclayton commented:
Updated to look more like HTML table

lavalamp commented:
OK, good enough for alpha. Still a little worried about that interface{}

lavalamp commented:
/lgtm
